### ISSUE TITLE
How do I implement embedded Java language in XML document (in a vscode extension)?

### ISSUE THREAD

**Issue by lispercat** [redhat-developer/vscode-java#1955]
I've read docs on https://code.visualstudio.com/api/language-extensions/embedded-languages but still wondering if it's possible to reuse vscode-java language server to provide intellisense/autocomplete features for Java/BeanShell code imbedded in XML like:

```<?xml version='1.0' encoding='UTF-8'?>
<!DOCTYPE Rule PUBLIC "sailpoint.dtd" "sailpoint.dtd">
<Rule language="beanshell" name="">
  <Description></Description>
<Source><![CDATA[
  private String concatenate(String s1, String s1){
    return s1 + s2;
  }
  return concatenate("foo", "bar");
]]></Source>
</Rule>
```

I guess I don't need to reimplement my own language server from scratch (though BeanShell is not strictly Java, but some reduced version of it) and can refer to the existing Java server from vscode-java. Any suggestions/samples?

**Comment by testforstephen** [comment]
The link you pasted above actually provides a sample [Language Server for Embedded Language with Language Services](https://github.com/microsoft/vscode-extension-samples/tree/main/lsp-embedded-language-service).

**Comment by angelozerr** [comment]
[Language Server for Embedded Language with Language Services](https://github.com/microsoft/vscode-extension-samples/tree/main/lsp-embedded-language-service) is only for HTML Language Service (which was written in TypeScript) not for XML. 

Th ething that I find it's shame with this sample is that it requires to start a new LSP client to manage your embedded language. It means that when an HTML file must be parsed, it will parsed twice : one by the HTML language server (extension of vscode) and one by the custom extension. IMHO I think HTML Language server should give the capability to extend by external component the custom embedded language. I mean https://github.com/microsoft/vscode-extension-samples/blob/main/lsp-embedded-language-service/server/src/embeddedSupport.ts should be in part of vscode-html-language-service and external component could contribute it.

As you want to manage embeded code inside XML, I suggest you that you work should be based on [vscode-xml](https://github.com/redhat-developer/vscode-xml) which gives the capability to extend any features (completion, validation, etc). See https://github.com/redhat-developer/vscode-xml/blob/master/docs/Extensions.md how to extend those features. 

vscode-xml is based on XML Language Server Lemminx written in Java. Perhaps lemMinx should provide an embedded support like https://github.com/microsoft/vscode-extension-samples/blob/main/lsp-embedded-language-service/server/src/embeddedSupport.ts. Please create an issue for that.

But the real question is `How to manage virtual document with JDT LS` ? From you sample,each `Source` XMLelement content should be managed as Java virtual document like https://github.com/microsoft/vscode-extension-samples/blob/f5e95b0e18899b0cf4779b5461b24c2511769472/lsp-embedded-language-service/server/src/modes/cssMode.ts#L21. 

I think JDT LS is working with real Java file and I'm not sure that you can open virtual document in JDT LS(ex : call a didOpen with a custom URI file which doesn't exists).

Don't hesitate to create issues on [LemMinx github project](https://github.com/eclipse/lemminx) for supporting embeded language if it is required (for the moment you could start to implement yourown XML completion participant and try to communicate with JDT LS with a virtual document). If embedded support can help, we could do that, but I repeat me, the first question is `How to manage virtual document with JDT LS`?

**Comment by lispercat** [comment]
Thanks so much Angelo for your response!
Since I know very little about LS, I think I'll need a plan to do some
small steps to finish the project.
As you suggested, I'll start by looking at vscode-xml, to see how it works
and how they can provide support for embedded languages.
Probably I also need to understand how JDT works, I think looking at
vscode-java should give me some clues on how you interact with JDT (to also
understand if JDT provides support to virtual documents).
BTW, when you say "manage virtual documents with JDT LS" do you mean
"embedded documents"? I am just trying to get more familiar with the
terminology.
Looks like a big project at the moment, but I think it's worth the effort.

**Comment by angelozerr** [comment]
Roland if you create your own compilation unit it means that you cannot
reuse the lsp editing completion validation etc and you must re implement
those features, right?

**Comment by rgrunber** [comment]
> Roland if you create your own compilation unit it means that you cannot reuse the lsp editing completion validation etc and you must re implement those features, right? Le mer. 9 juin 2021 à 20:24, Lispercat ***@***.***> a écrit :

I don't think so. Most of the completion/validation calls are based on the same ICompilationUnit interface so we should be able to reuse everything.

**Comment by angelozerr** [comment]
> I don't think so. Most of the completion/validation calls are based on the same ICompilationUnit interface so we should be able to reuse everything.

Great!

**Comment by lispercat** [comment]
> JDT-LS doesn't currently have the ability to create compilation units
that aren't based on a file, but I believe this could easily be done.

Do we need to propose a pull request to the JDT project for a non-file
based compilation unit?
Since yes, we would have to send a text that we extract from the embedded
region, so based on that text they would need to create a compilation unit.
Or maybe we can create a temp file with the embedded content and send it to
JDT as a compilation unit, but I guess there would be quite some overhead
if we do that.

**Comment by lispercat** [comment]
What I don't understand (after debugging some vscode-java + JDT) is that it
looks like the language server protocol doesn't provide support for
embedded documents (
https://microsoft.github.io/language-server-protocol/specifications/specification-current/
)
They assume that the whole content of the document specified by uri in the
RPC call will be parsed by the language server as a single compilation unit.
In this case the responsibility to recognize an embedded language falls on
the language server implementation.
But in this case the language server needs to know about some specific
embedded case.
The only way out (that I see now) is to:

   1. Extract embedded content in a separate file
   2. Send it to the language server
   3. Whatever positions the language server returns we need to translate
   them relative to the position of the embedded block

What do you think? Is there a better way or I have taken a completely wrong
approach?

**Comment by rgrunber** [comment]
Yup, that's exactly what I had in mind. The XML language server would be responsible for recognizing that there's embedded content and what other language server should handle it. It would then need to ensure all actions pertaining to that all line offsets get translated correctly to/from the embedded document.

I don't believe we need to make a PR into JDT for this. We could probably just have it in JDT-LS as there's plenty of utilities in both that shouldn't make it too difficult to implement compilation units.

**Comment by lispercat** [comment]
I wonder if we can propose a change to the language server API.
If we could supply range along with the url in the request, it could
automatically support embedded languages.
Now it looks like to compensate we'll need to artificially create a lot of
temp files in the background, but yes, then we could use JDT as is.

**Comment by lispercat** [comment]
Hi guys, can we resume the discussion?
I pretty much came full circle from trying to implement my own parser for
xml+embedded java to realizing that it would be much better to re-use JDT-LS
server only if it could use virtual documents.

At this point, it looks like when I register virtual document content
provider like so:
 vscode.workspace.registerTextDocumentContentProvider('embedded-content', {
     provideTextDocumentContent: uri => {
       .............
       return
"some-java-code-extracted-from-xml-where-xml-text-replaced-with-spaces";
}
});

and also registering provideCompletionItem in client middleware  so that it
returns CompletionList like this:

const completionList = await
vscode.commands.executeCommand<vscode.CompletionList>(
   'vscode.executeCompletionItemProvider',
        vdocUri,
        position,
       context.triggerCharacter
);

(I took inspiration from the lsp-embedded-request-forwarding sample (
https://github.com/microsoft/vscode-extension-samples/tree/main/lsp-embedded-request-forwarding
))

Looks like, even though java project is enabled and I can get correct
completions on any *.java files within the workspace folder, I see that

1) JDT-LS simply doesn't see the virtual document (even though the url
looks like this: `embedded-content://${original-xml-uri}.java` where
original-xml-uri is the path of xml file in the java project
2) Also I wonder if the "embedded-content" schema may remove the virtual
document from the "visibility scope" of JDT-LS which may only handle
documents with schema="file"

I think we are pretty close to getting the desirable support if we can
teach JDT-LS how to understand virtual documents and treat the content of
the file as part of the current project (using the current project's
Referenced Libraries).
What do you think guys, can I request the support for virtual documents in
JDB-LS?

**Comment by rgrunber** [comment]
The extension has its own [document content provider](https://github.com/redhat-developer/vscode-java/blob/master/src/providerDispatcher.ts#L28) and the language server has to be made aware of [how to deal with it](https://github.com/eclipse/eclipse.jdt.ls/blob/master/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JDTUtils.java#L424). Otherwise it likely ignores it.

I think you would need to write your own server side plugin that overrides the standard language server protocol methods and have it correctly handle your URIs.
